Consider file permissions when writing configuration in system tests

muhme opened the pull request:
Pull Request for Issue #43465.

### Summary of Changes
* Change `configuration.php` file mode to 644, write the file and restores the original file permissions in Cypress system test case `Installation.cy.js`
* updated system tests `README.md`

This change is developed for branch 4.4-dev and should be merged in all other branches.

### Testing Instructions
It was tested on new cloned 4.4-dev branch with the patch, running installation twice on macOS 14.4.1 (Sonoma), Windows 11 Pro and inside Docker container with Debian GNU/Linux 11 (bullseye).
* installation runs without errors == test passed
* `configuration.php` parameters are set and file mask is read-only for all (444)

### Actual result BEFORE applying this Pull Request
Installation fails on macOS and Windows


### Expected result AFTER applying this Pull Request
Installation step passed on macOS and Windows




### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

muhme committed:
Fix for issue #43465 writing configuration.php

Fix for issue #43465 'Cypress System Tests fail when writing configuration.php'
. remember the original file permission
. set 644
. write file
. restore original file permission

additional:
. writing file to ${Cypress.env('cmsPath')}/configuration.php` and no more to 'configuration.php'
. error handle file is not existing

muhme committed:
typo

muhme committed:
updated system tests README

laoneo commented on tests/System/README.md:
```diff
@@ -42,10 +60,12 @@ Tests should be:
 
-- **queryDB** Executes a query on the database
-- **cleanupDB** does some cleanup, is executed automatically after every test
+- **queryTestDB** executes a query on the database
```
I think this is wrong as queryTestDB is the internal function only, or?

muhme replied:
oh, yes, my fault, I was only looking in `db.js` at this moment and not into `index.js`-mapping, same for queryDB/cleanupDB - i will add one more commit

thx for your careful reading

richard67 commented on tests/System/README.md:
````diff
@@ -6,20 +6,38 @@ The CMS system tests are executed in real browsers and are using the [cypress.io
 
 1. Clone Joomla into a folder where it can be served by a web server
+```
+git clone --depth 1 https://github.com/muhme/joomla-cms 
````
I don’t think we should use your fork in this documentation.

muhme replied:
of course, thank you for checking, proposed change confirmed

muhme committed:
corrected task names

richard67 commented on tests/System/README.md:
````diff
@@ -6,20 +6,38 @@ The CMS system tests are executed in real browsers and are using the [cypress.io
 
 1. Clone Joomla into a folder where it can be served by a web server
+```
+git clone --depth 1 https://github.com/muhme/joomla-cms 
````
```suggestion
git clone --depth 1 https://github.com/joomla/joomla-cms 
```

muhme committed:
Update tests/System/README.md

of course, thank you for checking

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

laoneo commented on tests/System/support/commands/config.js:
```diff
@@ -1,18 +1,35 @@
+      const content = fileContent.replace(regex, `public $${parameter} = ${newValue};`);
+
+      // Remember the original file permissions
+      cy.task('getFilePermissions', configPath).then((originalPermissions) => {
```
I guess you can chain the return value of the task, so it is not so deep nested.

muhme replied:
I checked the code and found out that my inserted failure handler is not necessary, Cypress test fails with clear error message if the the file is not readable. So I deleted it and comitted a new version.

As I understand it, the four task's are already chained by promise and .then(). I do not know what to change, perhaps you want to show how your implementation would look like?

laoneo replied:
Something like this work too:
```
    // Remember the original file permissions
    cy.task('getFilePermissions', configPath).then((originalPermissions) => {
      // To be save, set write for owner and read for all
      cy.task('changeFilePermissions', { path: configPath, mode: '644' })
      // Write the changed file content back
      .then(() => cy.task('writeFile', { path: configPath, content }))
      // Restore the original file permissions
      .then(() => cy.task('changeFilePermissions', { path: configPath, mode: originalPermissions }));
    });
```

muhme replied:
yes, looks catchy, changed, thx 👍

muhme committed:
Merge branch 'joomla:4.4-dev' into fix-issue-43465

muhme committed:
deleted failure handler config_setParameter()

deleted failure handler for readFile as it is not needed, tested with chmod 0, Cypress fails with clear reason:

	CypressError: `cy.readFile("./configuration.php")` failed while trying to read the file at the following path:
	`.../43465/joomla-cms/configuration.php`
	The following error occurred:
	> "EACCES: permission denied, open '/Users/hlu/Desktop/no_backup/43465/joomla-cms/configuration.php'"

brianteeman commented on tests/System/plugins/fs.js:
```diff
@@ -33,4 +33,41 @@ function writeFile(path, content, config) {
+    return null;
+  } catch (err) {
+    console.error(`Failed to change file permissions: ${err}`);
+    // Rethrow to send the error to the Cypress tes
```
```suggestion
    // Rethrow to send the error to the Cypress test
```

brianteeman commented on tests/System/support/commands/config.js:
```diff
@@ -1,18 +1,26 @@
-    cy.task('writeFile', { path: 'configuration.php', content });
+    // Remember the original file permissions
+    cy.task('getFilePermissions', configPath).then((originalPermissions) => {
+      // To be save, set write for owner and read for all
```
```suggestion
      // To be safe, set write for owner and read for all
```

muhme replied:
Thank you for your corrections

muhme committed:
typo

Co-authored-by: Brian Teeman <brian@teeman.net>

muhme committed:
typo

Co-authored-by: Brian Teeman <brian@teeman.net>

muhme committed:
chain the then()-calls

Chaining the then()-calls for a not so deeply nested code source
looks catchy - thank Allon for the recommendation

laoneo commented on tests/System/support/commands/config.js:
```diff
@@ -1,18 +1,24 @@
+    // Remember the original file permissions
+    cy.task('getFilePermissions', configPath)
+      // To be save, set write for owner and read for all
+      .then((originalPermissions) => { cy.task('changeFilePermissions', { path: configPath, mode: '644' })
```
For better readability, I would put cy.task on a new line here and then intend the next thens one level as they do not belong to the chain with the forst then where you get the originalPermissions.

muhme replied:
mmh, The 1st version was
```JavaScript
// Remember the original file permissions
cy.task('getFilePermissions', configPath).then((originalPermissions) => {
  // To be save set read and write for owner
  cy.task('changeFilePermissions', { path: configPath, mode: '644' }).then(() => {
    // Write the changed file content back
    cy.task('writeFile', { path: configPath, content }).then(() => {
      // Restore the original file permissions
      cy.task('changeFilePermissions', { path: configPath, mode: originalPermissions });
    });
  });
});
```
actual 2nd version code formatting is:

```JavaScript
// Remember the original file permissions
cy.task('getFilePermissions', configPath)
  // To be save, set write for owner and read for all
  .then((originalPermissions) => { cy.task('changeFilePermissions', { path: configPath, mode: '644' })
  // Write the changed file content back
  .then(() => cy.task('writeFile', { path: configPath, content }))
  // Restore the original file permissions
  .then(() => cy.task('changeFilePermissions', { path: configPath, mode: originalPermissions }));
});
```

with your proposol, we end up in the following 3 version code, right?

```JavaScript
// Remember the original file permissions
cy.task('getFilePermissions', configPath)
  // To be save, set write for owner and read for all
  .then((originalPermissions) => {
  cy.task('changeFilePermissions', { path: configPath, mode: '644' })
    // Write the changed file content back
    .then(() => cy.task('writeFile', { path: configPath, content }))
    // Restore the original file permissions
    .then(() => cy.task('changeFilePermissions', { path: configPath, mode: originalPermissions }));
});
```

for me, as less expirienced JS programmer, the 1st version looks very clear, at first glance, there are four steps one after another, ok indent is three

the 2nd version starts at first glance with the first task and then continues with three `then()` and has the beauty of indent is only one - my favourite version

the 3rd formatting version doesn't looks intuitive for me that it's four steps, one after another; and indent is two, Or did I missunderstand your formatting proposal?

laoneo replied:
The second version is what I have proposed, but it looks not clear on the first glance because of the indentation. Why not like:

```
// Remember the original file permissions
    cy.task('getFilePermissions', configPath).then((originalPermissions) => {
      // To be save, set write for owner and read for all
      cy.task('changeFilePermissions', { path: configPath, mode: '644' })
        // Write the changed file content back
        .then(() => cy.task('writeFile', { path: configPath, content }))
        // Restore the original file permissions
        .then(() => cy.task('changeFilePermissions', { path: configPath, mode: originalPermissions }));
    });
```

Then you know that you first have fetched the file permissions because you need it in the last step. Afterwards you change it, then write and at the end restore as chain. For me it is more intuitive and simple than the nested ones. 2 and 3 are the same as my proposal, just with other intendation and line breaks.

muhme replied:
adopted code formatting for better readability as you proposed

laoneo replied:
Nice

muhme committed:
adopted code formatting for better readability

laoneo committed:
Merge branch '4.4-dev' into fix-issue-43465

laoneo commented:
There is a javascript error which prevents the system tests from running.

muhme committed:
fixing lint:js errors

- deleted console.log statements
- used const for never changing value
- refactored file mask to not use bitwise operation '&'

muhme commented:
> There is a javascript error which prevents the system tests from running.

fixed the lint:js errors and commited

muhme committed:
Merge branch '4.4-dev' into fix-issue-43465

alikon commented:
i was able to reproduce the issue on w10+laragon
![image](https://github.com/joomla/joomla-cms/assets/181681/6ece32af-353d-490e-9b5c-7bd3bed56cf9)
 
with pr

![image](https://github.com/joomla/joomla-cms/assets/181681/3b1dbaf7-fe70-4697-8461-37c3d1a51522)

alikon commented:
I have tested this item :white_check_mark: successfully on 3856d5b64339580ddc16818bbee31a50a78332d1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43466">issues.joomla.org/tracker/joomla-cms/43466</a>.</sub>

muhme committed:
fixed lint:testjs errors

muhme committed:
Merge branch 'fix-issue-43465' of https://github.com/muhme/joomla-cms into fix-issue-43465

laoneo committed:
Merge branch '4.4-dev' into fix-issue-43465

laoneo commented:
I would like to merge this one, but drone still fails. @muhme can you have a look? Looks like there is an issue with the API tests.

laoneo committed:
Merge branch '4.4-dev' into fix-issue-43465

muhme commented:
> I would like to merge this one, but drone still fails. @muhme can you have a look? Looks like there is an issue with the API tests.

I saw drone failed at the weekend, but I can't do anything with it, it looks unspecific with many failures, like the API failures. I tried to reproduce locally with fresh new clone, applied the PR, but running all system tests locally w/o problems.

At the moment there is a different picture, perhaps drone is started again? It doesn't come to phpmin-system-mysql step. I do not know the phan errors and do not know how to see npm errors in /tmp/npm-cache/_logs/2024-05-21T05_40_38_520Z-debug-0.log. sorry

alikon commented:
> I tried to reproduce locally with fresh new clone, applied the PR, but running all system tests locally w/o problems

same here too

laoneo commented:
You can check the artifacts, perhaps they give a clue.
![image](https://github.com/joomla/joomla-cms/assets/251072/5407c4c1-25c5-4fdd-8068-ce364550e60f)

muhme commented:
> You can check the artifacts, perhaps they give a clue.

checked some of the screenshot artifacts, just like with the detailed Cypress logs, unspecific with many errors, which in my opinion have nothing to do with this PR, unfortunately I can't help here, we need someone who can log into drone to check this

muhme committed:
Merge branch '4.4-dev' into fix-issue-43465

muhme commented:
I was able to force drone to run (by update PR with changes in the main branch) and i am wondering why the same 40 tests fail again. Since these are the API and mail related tests, the reason must be that setting the parameters in `configuration.php` does not work - I will check this again on Linux.

muhme commented:
I found one problem on (hardware based) Ubuntu for me:
* configuration.php is created by apache with user `www-data` and mode 444
* i am running cypress install as user `hlu` and i am therefore not allowed to change the file permission, nor to write the file

This is usually not a problem in Linux Docker containers, as the user is mostly `root` inside.

Therefore, the original problem occurs not only under macOS and Windows, but also under Linux as a non-root user.

Running the installation as `root` aka with `sudo` is possible with Cypress, but needs additional Cypress installation for `root`user and additional space of 653 MB installed in `root/cache/Cypress`. Should we change to sudo?
```
sudo npx cypress run --spec 'tests/System/integration/install/Installation.cy.js'
```

And it is not OK that the new `changeFilePermissions()` does not throw an error if it does not work, this should be able to be fixed.

And I think the same is for drone, `getFilePermissions()` or `changeFilePermissions()` fails silently and parameters in `configuration.php` are not configured. This is the needed to fix, but doesn't explain for me why it is not working on drone today and how is it working on drone without the PR.

muhme committed:
Better fix for configuration.php permission issue

Working with the code when fighting with the drone shows that a `chmod`
was already implemented in `writeFile()`. Following changes with this commit:
- Only using `chmod` method synchronously
- Replaced setting directory mode to setting file mode before writing
- Setting file mode only if the file exists
- Having final file mode as parameter with default 0o444
- Using 0o444 as default file mode and not hard-wired 0o777
- The methods `getFilePermissions()` and `changeFilePermissions()` created for this PR earlier are deleted.

Enhancement of the `tests/System/README.md` for troubleshooting three-user-problem in having
Cypress running user, web server running user and `root` user.

This commitment has been extensively tested in various combinations. Every test contains:
- Checking error before
- Doing the patch
- Running installation twice and running overall test suite

Tests are:
- macOS 14.5 Sonoma, local with apache & Cypress same user, branch 4.4-dev
  - error before `> EACCES: permission denied, open './configuration.php'`
- Docker, one container with joomla and one container with Cypress, using `root` users inside containers
  - no error before, but `configuration.php` is 777
  - after the patch `configuration.php` is 444 inside container and shown 644 on host
  - tested four times, branches 4.4-dev, 5.1-dev, 5.2-dev and 6.0-dev
- Ubuntu 24.04 LTS local installation, one non-root users running Cypress and
  another non-root user running Apache, branch 4.4-dev
  - error before `> EACCES: permission denied, open './configuration.php'`
  - need to use `sudo` and need to set `umask 0`, see troubleshooting
- Windows 11 Pro, Laragon with Cmder, branch 4.4-dev
  - error before `> EPERM: operation not permitted, open 'C:\laragon\www\joomla-cms\configuration.php'`

All tests are successful:
- running `Installation.cy.js` twice, checking `configuration.php` 444 and params are set
- running complete system test suite without errors

muhme committed:
configuration.php CMS path relative && umask 0

- corrected mistake task writeFile was used with cmsPath + 'configuration.php'
- extended writeFile to set process umask 0
  - to prevent the 3-user-problem == no need to set umask 0 in sudo anymore

This commitment has been tested in various combinations. Every test contains:
- Checking error before
- Doing the patch
- Running Installation.cy.js only and running overall test suite

Tests are:
- Docker environment with drone images, root running Cypress and www-data running Apache, branch 4.4-dev
  - no error before, but /tests/www/cmysql/configuration.php has 777
- Ubuntu 24.04 LTS local installation, one non-root user running Cypress and
  another non-root user running Apache, branch 4.4-dev
  - error before `> EACCES: permission denied, open './configuration.php'`
  - need to use `sudo`, see troubleshooting (umask 0 is no more needed)
- Windows 11 Pro, Laragon with Cmder, branch 4.4-dev
  - error before `> EPERM: operation not permitted, open 'C:\laragon\www\joomla-cms\configuration.php'`
  - found out that on the second run cy.exec('rm configuration.php') does not work under Windows
    - deleted file manually and i will create an issue afterwards to avoid enlarging this one
- macOS 14.5 Sonoma, local with apache & Cypress same user, branch 4.4-dev
  - error before `> EACCES: permission denied, open './configuration.php'`

All tests are successful:
- running `Installation.cy.js`, checking `configuration.php` 444 and params are set
- running complete system test suite without errors

laoneo commented:
Looks good now, restarted drone for another test run. If all is good, I'm goig to merge it.

laoneo commented:
Thank you very much, this one was a lot of work!
